src/System.Binary.Base64/System/Binary/Base64Decoder.cs

```diff
@@ -82,188 +67,120 @@ public static int ComputeDecodedLength(ReadOnlySpan<byte> source)
+            i0 <<= 18;
+            i1 <<= 12;
+            i2 <<= 6;
 
-            r0 = (byte)(i0 << 2 | i1 >> 4);
-            return true;
+            i0 |= i3;
+            i1 |= i2;
 
-            ErrorExit:
-            r0 = 0;
-            return false;
+            i0 |= i1;
+            return i0;
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private static int DecodePadByTwo(ref byte srcBytes)
+        private static void WriteThreeBytes(ref byte destBytes, int i0)
         {
-            byte b0 = srcBytes;
-            byte b1 = Unsafe.Add(ref srcBytes, 1);
-
-            if (!Decode(b0, b1, out byte r0)) return -1;
-
-            int result = r0;
-            return result;
+            destBytes = (byte)(i0 >> 16);
+            Unsafe.Add(ref destBytes, 1) = (byte)(i0 >> 8);
+            Unsafe.Add(ref destBytes, 2) = (byte)i0;
         }
 
         public TransformationStatus Transform(ReadOnlySpan<byte> source, Span<byte> destination, out int bytesConsumed, out int bytesWritten)
         {
             ref byte srcBytes = ref source.DangerousGetPinnableReference();
             ref byte destBytes = ref destination.DangerousGetPinnableReference();
 
-            int srcLength = source.Length / 4 * 4;  // only decode input up to the closest multiple of 4.
+            int srcLength = source.Length & ~0x3;  // only decode input up to the closest multiple of 4.
             int destLength = destination.Length;
 
             int sourceIndex = 0;
             int destIndex = 0;
 
             if (source.Length == 0) goto DoneExit;
 
-            int result = 0;
+            ref sbyte decodingMap = ref s_decodingMap[0];
 
             while (sourceIndex < srcLength - 4)
             {
-                result = Decode(ref Unsafe.Add(ref srcBytes, sourceIndex));
-                if (result == -1) goto InvalidExit;
+                int result = Decode(ref Unsafe.Add(ref srcBytes, sourceIndex), ref decodingMap);
+                if (result < 0) goto InvalidExit;
                 if (destIndex > destLength - 3) goto DestinationSmallExit;
-                Unsafe.WriteUnaligned(ref Unsafe.Add(ref destBytes, destIndex), result);
+                WriteThreeBytes(ref Unsafe.Add(ref destBytes, destIndex), result);
                 destIndex += 3;
                 sourceIndex += 4;
             }
 
             if (sourceIndex >= srcLength) goto NeedMoreExit;
 
-            int padding = 0;
+            int i0 = Unsafe.Add(ref srcBytes, srcLength - 4);
+            int i1 = Unsafe.Add(ref srcBytes, srcLength - 3);
+            int i2 = Unsafe.Add(ref srcBytes, srcLength - 2);
+            int i3 = Unsafe.Add(ref srcBytes, srcLength - 1);
+
+            i0 = Unsafe.Add(ref decodingMap, i0);
+            i1 = Unsafe.Add(ref decodingMap, i1);
 
-            if (Unsafe.Add(ref srcBytes, srcLength - 1) == s_encodingPad)
+            i0 <<= 18;
+            i1 <<= 12;
```

Comment:
I tried doing that, but i1 is different for some cases (it needs to get or'd with i2 first).

```C#
i0 |= i3;
i1 |= i2;

i0 |= i1;
```
vs. just
```C#
i0 |= i1;
```